storm-core/src/jvm/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -0,0 +1,252 @@
+        } catch (ClassNotFoundException e) {
+            LOG.error("Can't find blacklist strategy for name {}", strategyClassName);
+            throw new RuntimeException(e);
+        } catch (InstantiationException e) {
+            LOG.error("Throw InstantiationException blacklist strategy for name {}", strategyClassName);
+            throw new RuntimeException(e);
+        } catch (IllegalAccessException e) {
+            LOG.error("Throw illegalAccessException blacklist strategy for name {}", strategyClassName);
+            throw new RuntimeException(e);
+        }
+
+        nimbusMonitorFreqSecs = (Integer) _conf.get(Config.NIMBUS_MONITOR_FREQ_SECS);
+        blacklistStrategy.prepare(reporter, toleranceTime, toleranceCount, resumeTime, nimbusMonitorFreqSecs);
+
+        badSupervisorsTolerance = new CircularBuffer<HashMap<String, Set<Integer>>>(toleranceTime / nimbusMonitorFreqSecs);
+        cachedSupervisors = new HashMap<>();
+        blacklistHost = new HashSet<>();
+
+        StormMetricsRegistry.registerGauge("nimbus:num-blacklisted-supervisor", new Callable<Integer>() {
+            @Override
+            public Integer call() throws Exception {
+                //nimbus:num-blacklisted-supervisor + none blacklisted supervisor = nimbus:num-supervisors
+                return blacklistHost.size();
+            }
+        });
+    }
+
+    @Override
+    public void schedule(Topologies topologies, Cluster cluster) {
+        LOG.debug("running Black List scheduler");
+        Map<String, SupervisorDetails> supervisors = cluster.getSupervisors();
+        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+            SupervisorDetails supervisorDetails = entry.getValue();
+            String hosts = supervisorDetails.getHost();
+            Set<Integer> ports = supervisorDetails.getAllPorts();
+            LOG.debug("supervisor: {}, ports: {}", hosts, ports);
+        }
+        LOG.debug("AssignableSlots: {}", cluster.getAssignableSlots());
+        LOG.debug("AvailableSlots: {}", cluster.getAvailableSlots());
+        LOG.debug("UsedSlots: {}", cluster.getUsedSlots());
+
+        blacklistStrategy.resumeFromBlacklist();
+        badSupervisors(supervisors);
+        cluster.setBlacklistedHosts(getBlacklistHosts(cluster, topologies));
+        removeLongTimeDisappearFromCache();
+
+        underlyingScheduler.schedule(topologies, cluster);
+    }
+
+    private void badSupervisors(Map<String, SupervisorDetails> supervisors) {
+        Set<String> cachedSupervisorsKeySet = cachedSupervisors.keySet();
+        Set<String> supervisorsKeySet = supervisors.keySet();
+
+        Set<String> badSupervisorKeys = Sets.difference(cachedSupervisorsKeySet, supervisorsKeySet);//cached supervisor doesn't show up
+        HashMap<String, Set<Integer>> badSupervisors = new HashMap<String, Set<Integer>>();
+        for (String key : badSupervisorKeys) {
+            badSupervisors.put(key, cachedSupervisors.get(key));
```

Comment:
same as above.  this will only and bad supervisors but new ones. 
btw I think 'symmetricDifference' is what you mean.